src/common/scaled-scene-buffer.c

```diff
@@ -49,21 +49,63 @@ _cache_entry_destroy(struct scaled_scene_buffer_cache_entry *cache_entry, bool d
 	free(cache_entry);
 }
 
+static struct scaled_scene_buffer_cache_entry *
+find_cache_for_scale(struct scaled_scene_buffer *scene_buffer, double scale)
+{
+	struct scaled_scene_buffer_cache_entry *cache_entry;
+	wl_list_for_each(cache_entry, &scene_buffer->cache, link) {
+		if (cache_entry->scale == scale) {
+			return cache_entry;
+		}
+	}
+	return NULL;
+}
+
 static void
 _update_buffer(struct scaled_scene_buffer *self, double scale)
 {
 	self->active_scale = scale;
 
 	/* Search for cached buffer of specified scale */
-	struct scaled_scene_buffer_cache_entry *cache_entry, *cache_entry_tmp;
-	wl_list_for_each_safe(cache_entry, cache_entry_tmp, &self->cache, link) {
-		if (cache_entry->scale == scale) {
-			/* LRU cache, recently used in front */
-			wl_list_remove(&cache_entry->link);
-			wl_list_insert(&self->cache, &cache_entry->link);
-			wlr_scene_buffer_set_buffer(self->scene_buffer, cache_entry->buffer);
-			return;
+	struct scaled_scene_buffer_cache_entry *cache_entry =
+		find_cache_for_scale(self, scale);
+	if (cache_entry) {
+		/* LRU cache, recently used in front */
+		wl_list_remove(&cache_entry->link);
+		wl_list_insert(&self->cache, &cache_entry->link);
+		wlr_scene_buffer_set_buffer(self->scene_buffer, cache_entry->buffer);
+		/*
+		 * If found in our local cache,
+		 * - self->width and self->height are already set
+		 * - wlr_scene_buffer_set_dest_size() has already been called
+		 */
+		return;
+	}
+
+	/* Search from other cached scaled-scene-buffers */
+	struct scaled_scene_buffer *scene_buffer;
+	wl_list_for_each(scene_buffer, &self->impl->cached_buffers, link) {
```

Comment:
I struggle to get my head around this. Does each scaled-scene-buffer contain a list of 'other scaled scene buffers' in the impl-struct?

Reply:
Each scaled-scene-buffer has an impl pointer which in turn has a single list for that implementation (e.g. scaled-font-buffer) where all scene_buffers it has ever created (and their scene nodes are still alive) are stored. This would not change if we use a static wl_list per implementation and provide a `impl->get_cache()` getter for it. Do you have a suggestion to state that more clearly in the comment?

Reply:
Exactly. I should have added more detailed comments about it.